src/main/java/io/smartcat/ranger/core/DiscreteWeightedValue.java

```diff
@@ -0,0 +1,83 @@
+package io.smartcat.ranger.core;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+import org.apache.commons.math3.distribution.EnumeratedDistribution;
+import org.apache.commons.math3.util.Pair;
+
+/**
+ * Randomly selects one of the provided values following the provided weights.
+ *
+ * @param <T> Type this value would evaluate to.
+ */
+public class DiscreteWeightedValue<T> extends Value<T> {
```

Comment:
Instead of DiscreteWeightedValue, can we define something like WeightedDistribution? It could be constructed of a single list of probabilities (unnormalized). The random variable would be an index-type variable (integer, zero-based index). So the distribution sample (index) could be used to return discrete domain values.